### ISSUE TITLE
Mimic takes half a second to start up 

### ISSUE THREAD

**Issue by dylan-chong** [edgurgel/mimic#43]
this isn't a terribly big deal but at my company we've got a really slow start up time for tests. Currently it takes 13 seconds just to start up the test (not even run the test, not even compile either).

The biggest culprit is Prometheus, which is taking up 8-9 seconds of application start up time. Removing that library and stubbing out all it's modules, mimic comes in 2nd place at 481ms.

Wow 481ms? That's not a lot! :D but when it comes to having a fast TDD cycle, it's annoying to have to wait 4 seconds (even 13!) for unit test to start running. (Mind you, I'm also trying to figure out why the other libraries are also taking their time)

A way to show mimic takes half a second is by running

```
> time MIX_ENV=prod mix run
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1


real    0m0.937s
user    0m0.744s
sys     0m0.314s

Dylan@DylansWorkMBPM1:(main)~/Development/solve/mimic
> time MIX_ENV=prod mix run --no-start
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1


real    0m0.526s
user    0m0.330s
sys     0m0.319s
```

Both commands have compilation/IO overheads, the latter doesn't start up any applications. (prod env to avoid compiling dialyzer and such). The difference is about 400ms.

I'm curious really, why does it take almost half a second to start up Mimic? I understand that they're likely won't be any changes, but I'm curious than anything right now.

**Comment by edgurgel** [comment]
Hi @dylan-chong can you confirm which Mimic version is being used? We recently did some changes to speed this up https://github.com/edgurgel/mimic/releases/tag/v1.7.0

**Comment by edgurgel** [comment]
You can see some of the discussions we had recently https://github.com/edgurgel/mimic/pull/34 & https://github.com/edgurgel/mimic/pull/37

TL;DR We are now recompiling modules just when they are needed so it should speed up the time it takes to get to the first test.

**Comment by dylan-chong** [comment]
mimic 1.7.2

> elixir --version
Erlang/OTP 24 [erts-12.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Elixir 1.13.1 (compiled with Erlang/OTP 24)

Mac os 12.1, M1

**Comment by dylan-chong** [comment]
we have one instance of copy getting called `Mimic.copy(Sentry)` and that is all for the test i am running

**Comment by edgurgel** [comment]
Right. I assume `Mimic.copy` calls are all placed on `test_helper.exs` and not on the test itself?

**Comment by dylan-chong** [comment]
Correct

**Comment by edgurgel** [comment]
Looking again at this:

```
> time MIX_ENV=prod mix run
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1


real    0m0.937s
user    0m0.744s
sys     0m0.314s

Dylan@DylansWorkMBPM1:(main)~/Development/solve/mimic
> time MIX_ENV=prod mix run --no-start
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1


real    0m0.526s
user    0m0.330s
sys     0m0.319s
```

Why would Mimic be started on `prod` env? :thinking: 

Mimic should be included just on `test` and it doesn't do anything unless Mimic.copy is called which shouldn't help on a simple `mix run`

**Comment by edgurgel** [comment]
One way to avoid the price of starting the VM over and over is by using something like mix test_watch I think https://github.com/lpil/mix-test.watch/ ?

**Comment by dylan-chong** [comment]
same effect using `test` env.
I was benchmarking the `:application_controller.start_application/2` times and mimic was no 1 (after removing promethus) taking 400 ms. note doesn't recursively start apps and doesn't include complication/vm start times.
ooh thanks! didnt realise mix test watch doesnt restart the vm. we've been avoiding it due to it not working with umbrella apps, but might take a look at it again

**Comment by edgurgel** [comment]
TBH I'm not 100% sure if mix test watch does not restart the VM but I've used with umbrella apps without problems.

Thanks for the info I will have a look and see if there is anything we can do with the start-up speed

**Comment by dylan-chong** [comment]
> I've used with umbrella apps without problems.

oh, mix test watch does work at the top level of the umbrella! It looks like mix test watch has its own VM and spawns up a new VM to run tests each time. same slow issue sadly

**Comment by dylan-chong** [comment]
Been taking a look at how long mimic takes to start in my app

```
defmodule T do
  def t do
    {:ok, _} = Application.ensure_all_started(:rexbug)

    # Rexbug.start(":application_controller.start_application/2 :: return", time: 30_000, msgs: 1000, print_msec: true)
    # Rexbug.start("Mimic.Application.start/2 :: return", time: 30_000, msgs: 1000, print_msec: true)
    
    Application.ensure_loaded(:mimic)
    Mix.Tasks.Profile.Fprof.profile(&start/0, sort: :acc, warmup: false)
    # start()
  end

  defp start do
    apps = [:mimic]
    Enum.each(apps, fn app ->
      {:ok, _} = Application.ensure_all_started(app)
    end)
  end
end
```

```
MIX_ENV=test mix run --no-start -e "T.t"
```

Results

```
Total                                                              162     502.759       0.247
:fprof.apply_start_stop/4                                            0     502.759       0.005
T."-fun.start/0-"/0                                                  1     502.754       0.001
T.start/0                                                            1     502.753       0.001
Enum.each/2                                                          1     502.752       0.001
Enum."-each/2-lists^foreach/1-0-"/2                                  2     502.751       0.003
anonymous fn/1 in T.start/0                                          1     502.748       0.002
Application.ensure_all_started/1                                     1     502.746       0.001
Application.ensure_all_started/2                                     1     502.745       0.001
:application.ensure_all_started/2                                    1     502.744       0.003
:application.ensure_all_started/4                                   11     502.740       0.023
:application.ensure_started/4                                        2     502.696       0.010
:gen_server.call/3                                                  12     502.669       0.024
:gen.call/4                                                         12     502.645       0.012
:gen.do_for_proc/2                                                  12     502.633       0.024
anonymous fn/4 in :gen.call/4                                       12     502.597       0.012
:gen.do_call/4                                                      12     502.585       0.049
:application_controller.start_application/2                          2     502.527       0.002
:suspend                                                            13     502.512       0.000
:application_controller.is_running/1                                 8       0.124       0.008
:application.ensure_loaded/1                                         2       0.040       0.004
:application.load/1                                                  2       0.036       0.002
:application.load1/2                                                 2       0.034       0.004
:application_controller.load_application/1                           2       0.030       0.002
:application.get_key/2                                               4       0.016       0.004
:erlang.whereis/1                                                   12       0.012       0.012
:erlang.monitor/2                                                   12       0.012       0.012
:erlang.demonitor/2                                                 12       0.012       0.012
:application_controller.get_key/2                                    4       0.012       0.008
:ets.lookup/2                                                        4       0.004       0.004
:lists.reverse/1                                                     1       0.001       0.001
:undefined                                                           0       0.000       0.000
```

`:suspend` takes 600ms? 🤔 

With rexbug we confirm its mimic taking 1/2 second

```
# 15:41:21.373 #PID<0.93.0> :erlang.apply/2
# :application_controller.start_application(:tools, :temporary)

# 15:41:21.373 #PID<0.93.0> :erlang.apply/2
# :application_controller.start_application/2 -> :ok

# 15:41:21.373 #PID<0.93.0> :erlang.apply/2
# :application_controller.start_application(:mimic, :temporary)

# 15:41:21.982 #PID<0.93.0> :erlang.apply/2
# :application_controller.start_application/2 -> :ok                <-------------- ½ after prev msg
```

----------

Debugging this inside mimics project itself requires less hackiness

```diff
diff --git a/lib/mimic/application.ex b/lib/mimic/application.ex
index 12af5b4..27c7819 100644
--- a/lib/mimic/application.ex
+++ b/lib/mimic/application.ex
@@ -4,7 +4,9 @@ defmodule Mimic.Application do
   @moduledoc false
 
   def start(_, _) do
-    Cover.export_private_functions()
+    {micro, _} = :timer.tc(&Cover.export_private_functions/0)
+    IO.puts("TOOK AAAAAAAAAAAAAAAAA #{micro/1000} ms")
+
     children = [Server]
     Supervisor.start_link(children, name: Mimic.Supervisor, strategy: :one_for_one)
   end
diff --git a/mix.exs b/mix.exs
index fe769bc..5d88f05 100644
--- a/mix.exs
+++ b/mix.exs
@@ -33,7 +33,8 @@ defmodule Mimic.Mixfile do
   defp deps do
     [
       {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
-      {:credo, "~> 1.0", only: :dev}
+      {:credo, "~> 1.0", only: :dev},
+      {:rexbug, ">= 1.0.0"},
     ]
   end
 
diff --git a/mix.lock b/mix.lock
index 1b6ef6b..d6a92ae 100644
--- a/mix.lock
+++ b/mix.lock
@@ -9,4 +9,6 @@
   "makeup": {:hex, :makeup, "1.0.5", "d5a830bc42c9800ce07dd97fa94669dfb93d3bf5fcf6ea7a0c67b2e0e4a7f26c", [:mix], [{:nimble_parsec, "~> 0.5 or ~> 1.0", [hex: :nimble_parsec, repo: "hexpm", optional: false]}], "hexpm", "cfa158c02d3f5c0c665d0af11512fed3fba0144cf1aadee0f2ce17747fba2ca9"},
   "makeup_elixir": {:hex, :makeup_elixir, "0.15.0", "98312c9f0d3730fde4049985a1105da5155bfe5c11e47bdc7406d88e01e4219b", [:mix], [{:makeup, "~> 1.0", [hex: :makeup, repo: "hexpm", optional: false]}, {:nimble_parsec, "~> 1.1", [hex: :nimble_parsec, repo: "hexpm", optional: false]}], "hexpm", "75ffa34ab1056b7e24844c90bfc62aaf6f3a37a15faa76b07bc5eba27e4a8b4a"},
   "nimble_parsec": {:hex, :nimble_parsec, "1.1.0", "3a6fca1550363552e54c216debb6a9e95bd8d32348938e13de5eda962c0d7f89", [:mix], [], "hexpm", "08eb32d66b706e913ff748f11694b17981c0b04a33ef470e33e11b3d3ac8f54b"},
+  "redbug": {:hex, :redbug, "1.2.2", "366d8961770ddc7bb5d209fbadddfa7271005487f938c087a0e385a57abfee33", [:rebar3], [], "hexpm", "b5fe7b94e487be559cb0ec1c0e938c9761205d3e91a96bf263bdf1beaebea729"},
+  "rexbug": {:hex, :rexbug, "1.0.5", "e4fce59d1cb4f574b2d84181507b4782bc4b6afcb64e2cd276003c563ffef766", [:mix], [{:mix_test_watch, ">= 0.5.0", [hex: :mix_test_watch, repo: "hexpm", optional: true]}, {:redbug, "~> 1.2", [hex: :redbug, repo: "hexpm", optional: false]}], "hexpm", "13a3f180a9e490686a774725a07a21caf05735b7c012824d86960e9541aab46a"},
 }
```

```
> mix run
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1

Compiling 1 file (.ex)
TOOK AAAAAAAAAAAAAAAAA 455.647 ms
```

```
diff --git a/lib/mimic/cover.ex b/lib/mimic/cover.ex
index 53b4d16..330d1fb 100644
--- a/lib/mimic/cover.ex
+++ b/lib/mimic/cover.ex
@@ -8,10 +8,16 @@ defmodule Mimic.Cover do
 
   @doc false
   def export_private_functions do
-    {_, binary, _} = :code.get_object_code(:cover)
-    {:ok, {_, [{_, {_, abstract_code}}]}} = :beam_lib.chunks(binary, [:abstract_code])
-    {:ok, module, binary} = :compile.forms(abstract_code, [:export_all])
-    :code.load_binary(module, '', binary)
+    {_, binary, _}                        = time fn -> :code.get_object_code(:cover) end
+    {:ok, {_, [{_, {_, abstract_code}}]}} = time fn -> :beam_lib.chunks(binary, [:abstract_code]) end
+    {:ok, module, binary}                 = time fn -> :compile.forms(abstract_code, [:export_all]) end
+    time fn -> :code.load_binary(module, '', binary) end
+  end
+
+  defp time(fun) do
+    {micro, res} = :timer.tc(fun)
+    IO.puts("TOOK AAAAAAAAAAAAAAAAA #{micro/1000} ms")
+    res
   end
 ```
 
 ```
 > mix run
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1

TOOK AAAAAAAAAAAAAAAAA 0.313 ms
TOOK AAAAAAAAAAAAAAAAA 5.136 ms
TOOK AAAAAAAAAAAAAAAAA 450.099 ms
TOOK AAAAAAAAAAAAAAAAA 1.645 ms
 ```
 
 so it looks like `:compile.forms(abstract_code, [:export_all])` is whats taking the ½ second

**Comment by dylan-chong** [comment]
Fprof really didnt like profiling `:beam_lib.chunks(binary, [:abstract_code])`. It ate 17GB of memory and ran for over an hour until i decided to control c it. bizarre

**Comment by dylan-chong** [comment]
So why would this function be so slow?
And is it possible to be run it as a Task under the app supervisor so it runs async to any other setup happening in the user applications?

**Comment by edgurgel** [comment]
Nice finding. We need to recompile `cover` so that Mimic can collect coverage data and it has to happen before a single test starts.

What we can avoid is doing this when we are not actually collecting coverage (which is most of the time). One way would be when [we check if the module is cover enabled](https://github.com/edgurgel/mimic/blob/60e73c8805688e3ec643818c5ee248f74c74e928/lib/mimic/cover.ex#L9-L12) we could then recompile cover if it hasn't been recompiled yet. One way to know if the cover recompilation has already happened is by doing `function_exported?(:cover, :get_term, 1)`.

**Comment by dylan-chong** [comment]
Ooh, I’ll give it a go and see if that works!

**Comment by dylan-chong** [comment]
It's working!

```
> time mix test --max-failures 1
..........................................................................

Finished in 0.7 seconds (0.7s async, 0.00s sync)
74 tests, 0 failures

Randomized with seed 65079

real    0m1.349s
user    0m0.721s
sys     0m0.463s

> time mix test --cover --max-failures 1
..........................................................................

Finished in 0.7 seconds (0.7s async, 0.00s sync)
74 tests, 0 failures

Randomized with seed 367421

real    0m2.054s
user    0m1.648s
sys     0m0.466s
```
